### ISSUE TITLE
Efficiently reading partitions of a partitioned Parquet file using Dask

### ISSUE THREAD

**Issue by andrethrill** [dask/dask#3571]
This is a bit of a continuation from #3168 

You can jump directly to the section **The Problem**. Or read through the **Aux code as well**

<details>
<summary>Aux code</summary>

# Imports

```python
import numpy as np
import pandas as pd
import dask.dataframe as dd
import random
import fastparquet
import pprint
```

# Generation and plotting functions

```python
#generate dummy df
def generate_df(size):
    d = {
        'dates':
        pd.date_range('1901-01-01', periods=size, freq='1s'),
        'signal1':
        np.random.normal(0, 0.3, size=size).cumsum() + 50,
        'signal2':
        np.random.normal(0, 0.3, size=size).cumsum() + 50,
        'signal3':
        np.random.normal(0, 0.3, size=size).cumsum() + 50,
        'fake_categorical1':
        np.random.choice(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I'], size=size),
        'fake_categorical2':
        np.random.choice(700, size=size)
    }

    df = pd.DataFrame(d)

    return df.melt(id_vars=['dates', 'fake_categorical1', 'fake_categorical2'])
```

# Generate df

Let's generate a Dataframe:

```python
# fake_categorical2 purposefully has a big number of different possible values.
# This is similar to the use case I currently have with my real dataset.
# For each value of fake_categorical2, my real dataset has a big number of different values.

Hence the need to partioning it.
big_df = generate_df(1000*1000*50)
big_df.head(30)
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dates</th>
      <th>fake_categorical1</th>
      <th>fake_categorical2</th>
      <th>variable</th>
      <th>value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1901-01-01 00:00:00</td>
      <td>B</td>
      <td>662</td>
      <td>signal1</td>
      <td>49.944626</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1901-01-01 00:00:01</td>
      <td>G</td>
      <td>306</td>
      <td>signal1</td>
      <td>50.160813</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1901-01-01 00:00:02</td>
      <td>H</td>
      <td>235</td>
      <td>signal1</td>
      <td>50.149038</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1901-01-01 00:00:03</td>
      <td>G</td>
      <td>573</td>
      <td>signal1</td>
      <td>50.399590</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1901-01-01 00:00:04</td>
      <td>C</td>
      <td>433</td>
      <td>signal1</td>
      <td>50.203727</td>
    </tr>
    <tr>
      <th>5</th>
      <td>1901-01-01 00:00:05</td>
      <td>C</td>
      <td>355</td>
      <td>signal1</td>
      <td>49.834526</td>
    </tr>
    <tr>
      <th>6</th>
      <td>1901-01-01 00:00:06</td>
      <td>G</td>
      <td>504</td>
      <td>signal1</td>
      <td>49.635185</td>
    </tr>
    <tr>
      <th>7</th>
      <td>1901-01-01 00:00:07</td>
      <td>A</td>
      <td>656</td>
      <td>signal1</td>
      <td>49.702752</td>
    </tr>
    <tr>
      <th>8</th>
      <td>1901-01-01 00:00:08</td>
      <td>C</td>
      <td>146</td>
      <td>signal1</td>
      <td>49.379890</td>
    </tr>
    <tr>
      <th>9</th>
      <td>1901-01-01 00:00:09</td>
      <td>H</td>
      <td>129</td>
      <td>signal1</td>
      <td>48.768839</td>
    </tr>
    <tr>
      <th>10</th>
      <td>1901-01-01 00:00:10</td>
      <td>B</td>
      <td>279</td>
      <td>signal1</td>
      <td>48.866203</td>
    </tr>
    <tr>
      <th>11</th>
      <td>1901-01-01 00:00:11</td>
      <td>F</td>
      <td>58</td>
      <td>signal1</td>
      <td>49.389052</td>
    </tr>
    <tr>
      <th>12</th>
      <td>1901-01-01 00:00:12</td>
      <td>H</td>
      <td>467</td>
      <td>signal1</td>
      <td>49.377938</td>
    </tr>
    <tr>
      <th>13</th>
      <td>1901-01-01 00:00:13</td>
      <td>F</td>
      <td>403</td>
      <td>signal1</td>
      <td>49.841742</td>
    </tr>
    <tr>
      <th>14</th>
      <td>1901-01-01 00:00:14</td>
      <td>G</td>
      <td>463</td>
      <td>signal1</td>
      <td>49.747901</td>
    </tr>
    <tr>
      <th>15</th>
      <td>1901-01-01 00:00:15</td>
      <td>G</td>
      <td>539</td>
      <td>signal1</td>
      <td>50.245040</td>
    </tr>
    <tr>
      <th>16</th>
      <td>1901-01-01 00:00:16</td>
      <td>H</td>
      <td>558</td>
      <td>signal1</td>
      <td>50.447821</td>
    </tr>
    <tr>
      <th>17</th>
      <td>1901-01-01 00:00:17</td>
      <td>H</td>
      <td>256</td>
      <td>signal1</td>
      <td>50.740859</td>
    </tr>
    <tr>
      <th>18</th>
      <td>1901-01-01 00:00:18</td>
      <td>A</td>
      <td>98</td>
      <td>signal1</td>
      <td>51.025387</td>
    </tr>
    <tr>
      <th>19</th>
      <td>1901-01-01 00:00:19</td>
      <td>B</td>
      <td>333</td>
      <td>signal1</td>
      <td>51.141644</td>
    </tr>
    <tr>
      <th>20</th>
      <td>1901-01-01 00:00:20</td>
      <td>A</td>
      <td>111</td>
      <td>signal1</td>
      <td>51.178237</td>
    </tr>
    <tr>
      <th>21</th>
      <td>1901-01-01 00:00:21</td>
      <td>A</td>
      <td>637</td>
      <td>signal1</td>
      <td>50.929670</td>
    </tr>
    <tr>
      <th>22</th>
      <td>1901-01-01 00:00:22</td>
      <td>F</td>
      <td>187</td>
      <td>signal1</td>
      <td>51.014629</td>
    </tr>
    <tr>
      <th>23</th>
      <td>1901-01-01 00:00:23</td>
      <td>G</td>
      <td>72</td>
      <td>signal1</td>
      <td>51.108230</td>
    </tr>
    <tr>
      <th>24</th>
      <td>1901-01-01 00:00:24</td>
      <td>B</td>
      <td>418</td>
      <td>signal1</td>
      <td>50.877421</td>
    </tr>
    <tr>
      <th>25</th>
      <td>1901-01-01 00:00:25</td>
      <td>E</td>
      <td>530</td>
      <td>signal1</td>
      <td>51.534064</td>
    </tr>
    <tr>
      <th>26</th>
      <td>1901-01-01 00:00:26</td>
      <td>B</td>
      <td>379</td>
      <td>signal1</td>
      <td>51.539611</td>
    </tr>
    <tr>
      <th>27</th>
      <td>1901-01-01 00:00:27</td>
      <td>F</td>
      <td>389</td>
      <td>signal1</td>
      <td>51.367636</td>
    </tr>
    <tr>
      <th>28</th>
      <td>1901-01-01 00:00:28</td>
      <td>I</td>
      <td>566</td>
      <td>signal1</td>
      <td>51.374062</td>
    </tr>
    <tr>
      <th>29</th>
      <td>1901-01-01 00:00:29</td>
      <td>H</td>
      <td>6</td>
      <td>signal1</td>
      <td>51.305901</td>
    </tr>
  </tbody>
</table>
</div>

# Save partitioned parquet

Save a **partitioned** parquet:

```python
fastparquet.write('file_partitioned', big_df, compression='snappy',
                  file_scheme='hive', write_index=False,
                  partition_on=['fake_categorical1','fake_categorical2'])
```

</details>

# The Problem

## Read it with fastparquet

Running `ParquetFile` takes ~900ms. I guess this time is due to size of the parquet `metadata`

```python
%timeit pf = fastparquet.ParquetFile('file_partitioned')
```

    893 ms ± 203 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Getting the filtered dataframe takes ~200ms

```python
%timeit parquet_df = pf.to_pandas(filters=[('fake_categorical1', '==', 'F'), ('fake_categorical2', '==', 418)])
```

    201 ms ± 7.09 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

This means that *after* running `ParquetFile`, if I want to run computations on different partitions I can load *each* filtered dataframe in ~200ms **avoiding the initial overhead of ~900ms**

## Read it with Dask

Getting the filtered dataframe by using the `filters` argument takes ~1.31s (a bit more than the sum of 900ms overhead + 200ms above)

```python
%timeit dsk_1_df = dd.read_parquet('file_partitioned', filters=[('fake_categorical1', '==', 'F'), ('fake_categorical2', '==', 418)]).compute()
```

    1.31 s ± 251 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

I suspect this is because, internally, Dask is running the same, or equivalent, to `ParquetFile` above.

I hoped  that I could run the part equivalent to `ParquetFile` only once and then be able to get different filtered dataframes with at least the same performance as with fastparquet above (in ~200ms without the 900ms overhead).

So I tried it like this:

1 - `read_parquet` without the `filters` argument (this could run the equivalent to `ParquetFile` above just once, reading all the metadata). And takes ~1.1s

```python
%timeit dsk_2 = dd.read_parquet('file_partitioned')
```

    1.07 s ± 117 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

2 - I was hoping that after when I do a `loc` on a column that was used to partition the parquet file, Dask would be aware of that and would simplify the computation graph and avoid loading all partitions, but from the time that the line below is taking I can see that this is not the case. (I've waited for 30 min and then cancelled it)

```python
%timeit dsk_2_df = dsk_2.loc[(dsk_2.fake_categorical1=='F')&(dsk_2.fake_categorical2==418)].compute()
```

# Question

1. Is there away to, in Dask,  run the initial overhead of loading the parquet `metadata` just once and then be able to get filtered dataframes that will only "touch" the relevant partitions?

Thanks in advance!

**Comment by martindurant** [comment]
Unfortunately, Dask does not handle `dsk_2.loc[conditions]` to "push down" filters to fastparquet, except in the simplest case of the single index column. In fact this is complex to do, the dask dataframe does not maintain column statistics except for the index divisions, and don't maintain a reference to the original parquet file (the graph contains only the read-row-group method calls).

I agree that it ought to be possible to construct a dask dataframe as

```
df = dd.read_parquet(pf, filters=...)
```
where `pf` is the ParquetFile instance, so that you don't need to reread the metadata. This would be easy to implement.

I wonder why this particular dataset takes so long to read the metadata?

**Comment by martindurant** [comment]
From your code, am I right in guessing that you have *very many* partitions, but each is very small? That is somewhat of an anti-pattern for parquet storage.

**Comment by andrethrill** [comment]
@martindurant for the example that I posted here (using the dummy dataframe) you are right that each partition is very small. But, with my real dataset that is not the case (real dataset is ~50GB on disk with snappy compression, whereas this one is 2GB). 




But I do like your suggestion of implementing `df = dd.read_parquet(pf, filters=...)` I think this would be perfect for my use case. I can give it a try over the rest of this week (can't promise I'll succeed though :) ).

**Comment by martindurant** [comment]
It will take a little digging, I think, since the dask code path passes around filesystem objects, whereas the ParquetFile instance only keeps the reference to the `open` method. Some of the options to `read_parquet` would be incompatible with passing an instance, I think `storage_options` and `engine` (would implicitly be fastparquet).

By the way, would be interesting to see if/how much faster pyarrow is at interpreting the metadata for your case. Of course, it doesn't have the equivalent of `filters=` (yet?), so it may not be a viable solution.

**Comment by andrethrill** [comment]
@martindurant please have a look at #3573 and tell me what you think

**Comment by andrethrill** [comment]
Solved by #3573
